README
This is BSD, a package for interfacing to various BSD UNIX system calls
and library routines.

This package is a freely available open source package under the "Berkeley"
license, same as Tcl.  You can do virtually anything you like with it, such as 
modifying it, redistributing it, and selling it either in whole or in part.  
See the file "license.terms" for complete information.

Tweezer was written by Karl Lehenbauer.

USING THE BSD INTERFACE
=======================

package require BSD


BUGS
====


CONTENTS
========

Makefile.in	Makefile template.  The configure script uses this file to
		produce the final Makefile.

README		This file

aclocal.m4	Generated file.  Do not edit.  Autoconf uses this as input
		when generating the final configure script.  See "tcl.m4"
		below.

configure	Generated file.  Do not edit.  This must be regenerated
		anytime configure.in or tclconfig/tcl.m4 changes.

configure.in	Configure script template.  Autoconf uses this file as input
		to produce the final configure script.

generic/bsd.c	Tcl BSD interface source code
generic/tclbsd.c

generic/bsd.h	Header files for the BSD interface.


tclconfig/	This directory contains various template files that build
		the configure script.  They should not need modification.

	install-sh	Program used for copying binaries and script files
			to their install locations.

	tcl.m4		Collection of Tcl autoconf macros.  Included by
			aclocal.m4 to define SC_* macros.

UNIX BUILD
==========

Building under most UNIX systems is easy, just run the configure script
and then run make. 

	$ cd tclbsd
	$ ./configure
	$ make
	$ make install

WINDOWS BUILD
=============

Probably irrelevant

INSTALLATION
============

The BSD interface package installs like so:

         $exec_prefix
          /       \
        lib       bin
         |         |
   PACKAGEx.y   (dependent .dll files on Windows)
         |
  pkgIndex.tcl (.so|.dll files)

The main .so|.dll library file gets installed in the versioned PACKAGE
directory, which is OK on all platforms because it will be directly
referenced with by 'load' in the pkgIndex.tcl file.  Dependent DLL files on
Windows must go in the bin directory (or other directory on the user's
PATH) in order for them to be found.



ChangeLog
2011-08-27 karl
    * Add "gjournal" to flags potentially returned by statfs and getfsstat 
      indicating that journaling with gjournal is enabled (see gjournal(8)).

2011-02-14 karl
    * "rlimit virtual" now works -- a break statement was missing.

    * Silence "clang --analyze" warnings.

2011-01-18 karl
    * Version bump to 1.4.  Tagged 1.4 release.

2010-08-12 karl
    * Upgrade to Tcl Extension Architecture (TEA) 3.8.

2010-03-22 karl
    * Upgrade to Tcl Extension Architecture (TEA) 3.7.

2009-08-11 karl
    * Add new command, cptime, that uses the sysctl library call to get
      CPU time statistics.  Command takes an array name as an argument
      and will fill it with system, user, nice, interrupt, idle, and
      total clock counts.  Upon a second invocation with the same array
      as an argument, it will calculate the cpu time of the various
      elements as a percentage, returning those as key-value pairs and
      update the array with the latest sampled values.

    * Version bump to 1.3.

2009-07-28 karl
    * Update README.FreeBSD to reflect that --with-tclinclude is not needed
      and --with-mandir will put the manpage in a place where "man" will
      find it without changing MANPATH.  Update .gitignore to ignore *.o *.so.

2009-03-31 karl
    * Changes to make tclbsd work on Debian Linux.

2008-06-23 karl
    * sockbuf is not available on Mac OS X Leopard.  Have it be an error
      to try to get or set that value.

    * OS X uses termios rather than termio so fix the Makefile so it will
      build properly on OS X.

2008-05-27 karl
    * Add new rlimit-settable values for max virtual memory mapped ("virtual")
      and max socket buffer memory (mbufs, in bytes) ("sockbuf").

    * You can now specify "unlimited" as an argument to rlimit value setting
      options and tclbsd will substitute RLIM_INFINITY as the value.

    * Switched the data type of the variable that we use for limits from
      long to Tcl wide int, because rlim_t type is a 64 bit integer.

    * Got rid of intermediate value generation with sprintf -- just
      using standard Tcl functions like Tcl_NewWideIntObj instead.

    * Checking error returns in a couple places where we weren't.

2008-04-15 karl
    * Add ::bsd::abort command to invoke the abort() system call.

2007-12-21 karl
    * Fold helper procs from tclbsd.tcl into bsd_procs.tcl.

    * Version bump to 1.1.

2007-06-13 karl
    * The ident to openlog must be a const char *, i.e. it needs to not
      change after the openlog, so use Tcl_IncrRefCount on the Tcl object
      containing the ident to prevent the Tcl interpreter from later
      taking the space for some other purpose.

2007-06-12 karl
    * Document the ::bsd::syslog command.

    * Emit reasonable error messages when the facility or priority is
      unrecognized.

2007-03-12 karl
    * Update to Tcl Extension Architecture (TEA) 3.6.

2007-02-27 karl
    * Made getfsstat actually work.

    * Support the -wait and -nowait options to getfsstat, which is a Tcl
      interface to the BSD UNIX "getfsstat" system call.

    * Version bump to 1.0.

2007-02-18 karl
    * Add getfsstat that will return filesystem statistics for all mounted
      filesystems.

2007-01-17 karl
    * Add cputime proc to calculate the CPU time used by running something,
      like Tcl's time but uses the CPU time not elapsed real time.

2006-01-04 Karl Lehenbauer <karl-bsd@sc.com>
    * Now that we're loading in Tcl procs as well as a C extension
      library, make the generation of the pkgIndex.tcl file be
      correct, eh?

2006-01-04 Karl Lehenbauer <karl-bsd@sc.com>
    * Add support for syslog -- yes there are other Tcl syslog interfaces,
      at least Scotty/Tnm, but we don't want to pull all that stuff in,
      and this interface is, we believe, BSD-complete.

    * configure.in, generic/bsd.c: Detect whether the setproctitle call
      is available so we can build on the BSD variant Mac OS X/Darwin.

2005-11-28 Karl Lehenbauer <karl-bsd@sc.com>
    * Add support for ::bsd::getkey to read a single key from the keybaord
      in raw mode and return the character value as an integer

2005-11-23 Karl Lehenbauer <karl-bsd@sc.com>
    * Add support for MNT_JAILDEVFS and MNT_IGNORE flags when present.

2005-11-28 Karl Lehenbauer <karl-bsd@sc.com>
    * Add support for ::bsd::getkey to read a single key from the keybaord
      in raw mode and return the character value as an integer

2005-11-23 Karl Lehenbauer <karl-bsd@sc.com>
    * Add support for MNT_JAILDEVFS and MNT_IGNORE flags when present.

2005-08-07 Karl Lehenbauer <karl-bsd@sc.com>

    * Put all commands in a "bsd" namespace
    * Export commands from "bsd" namespace

generic/bsd.c

/* 
 * bsd.c --
 *
 *      Contains the Superconnect BSD extensions for Tcl.
 *-----------------------------------------------------------------------------
 * Copyright 1998 NeoSoft, 2005 Superconnect.
 *
 * Permission to use, copy, modify, and distribute this software and its
 * documentation for any purpose and without fee is hereby granted, provided
 * that the above copyright notice appear in all copies.  NeoSoft makes
 * no representations about the suitability of this software for any purpose.
 * It is provided "as is" without express or implied warranty.
 *
 *-----------------------------------------------------------------------------
 * $Id: bsd.c,v 1.13 2008-06-23 16:07:33 karl Exp $
 *-----------------------------------------------------------------------------
 */

#include <sys/types.h>
#include <unistd.h>
#include <sys/time.h>
#include <sys/resource.h>

#include <sys/param.h>
#include <sys/mount.h>

#ifdef HAVE_SYS_VFS_H
#  include <sys/vfs.h>
#endif

#include <stdlib.h>
#include <string.h>

#include <tcl.h>

static int
AppendNameLong (interp, listObj, name, element)
    Tcl_Interp    *interp;
    Tcl_Obj       *listObj;
    char          *name;
    long           element;
{
    Tcl_Obj   *stringObj;
    Tcl_Obj   *elementObj;

    stringObj = Tcl_NewStringObj (name, -1);
    if (Tcl_ListObjAppendElement (interp, listObj, stringObj) != TCL_OK) {
	return TCL_ERROR;
    }

    elementObj = Tcl_NewLongObj (element);
    if (Tcl_ListObjAppendElement (interp, listObj, elementObj) != TCL_OK) {
	return TCL_ERROR;
    }
    return TCL_OK;
}

static int
AppendNameDouble (interp, listObj, name, element)
    Tcl_Interp    *interp;
    Tcl_Obj       *listObj;
    char          *name;
    double         element;
{
    Tcl_Obj   *stringObj;
    Tcl_Obj   *elementObj;

    stringObj = Tcl_NewStringObj (name, -1);
    if (Tcl_ListObjAppendElement (interp, listObj, stringObj) != TCL_OK) {
	return TCL_ERROR;
    }

    elementObj = Tcl_NewDoubleObj (element);
    if (Tcl_ListObjAppendElement (interp, listObj, elementObj) != TCL_OK) {
	return TCL_ERROR;
    }

    return TCL_OK;
}

static int
AppendNameString (interp, listObj, name, element)
    Tcl_Interp    *interp;
    Tcl_Obj       *listObj;
    char          *name;
    char          *element;
{
    Tcl_Obj   *stringObj;
    Tcl_Obj   *elementObj;

    stringObj = Tcl_NewStringObj (name, -1);
    if (Tcl_ListObjAppendElement (interp, listObj, stringObj) != TCL_OK) {
	return TCL_ERROR;
    }

    elementObj = Tcl_NewStringObj (element, -1);
    if (Tcl_ListObjAppendElement (interp, listObj, elementObj) != TCL_OK) {
	return TCL_ERROR;
    }
    return TCL_OK;
}

static int
AppendNameObj (interp, listObj, name, appendListObj)
    Tcl_Interp    *interp;
    Tcl_Obj       *listObj;
    char          *name;
    Tcl_Obj       *appendListObj;
{
    Tcl_Obj   *stringObj;

    stringObj = Tcl_NewStringObj (name, -1);
    if (Tcl_ListObjAppendElement (interp, listObj, stringObj) != TCL_OK) {
	return TCL_ERROR;
    }

    if (Tcl_ListObjAppendElement (interp, listObj, appendListObj) != TCL_OK) {
	return TCL_ERROR;
    }
    return TCL_OK;
}

/*-----------------------------------------------------------------------------
 * BSD_RusageObjCmd --
 *  
 * Implements the `rusage' command:
 *    rusage self
 *    rusage children
 *  
 * Results:
 *      A standard Tcl result.
 *      
 * Side effects:
 *      See the user documentation.
 *-----------------------------------------------------------------------------
 */     
int
BSD_RusageObjCmd (clientData, interp, objc, objv)
    ClientData    clientData;
    Tcl_Interp   *interp;
    int           objc;
    Tcl_Obj      *CONST objv[];
{
    char         *subOption = NULL;
    Tcl_Obj      *resultObj = Tcl_GetObjResult (interp);
    int           who;
    struct rusage rusage;
    double	  utime;
    double	  stime;
    double        ixrssSeconds;
    double        idrssSeconds;
    double        isrssSeconds;

    static int    clockTick = -1;

    if (clockTick == -1) {
	clockTick = sysconf(_SC_CLK_TCK);
    }

    if (objc > 2) {
	Tcl_WrongNumArgs (interp, 1, objv, "[self|children]");
	return TCL_ERROR;
    }

    if (objc == 2) {
	subOption = Tcl_GetStringFromObj (objv[1], NULL);
    }

    if (objc == 1 || strcmp (subOption, "self") == 0) {
	who = RUSAGE_SELF;
    } else if (strcmp (subOption, "children") == 0) {
	who = RUSAGE_CHILDREN;
    } else {
	Tcl_AppendStringsToObj (resultObj, 
				"subcommand was not \"self\" or \"children\"",
				(char *)NULL);
	return TCL_ERROR;
    }

    if (getrusage (who, &rusage) < 0) {
	Tcl_SetStringObj (resultObj, Tcl_PosixError (interp), -1);
	return TCL_ERROR;
    }

    utime = 
        rusage.ru_utime.tv_sec + rusage.ru_utime.tv_usec / 1000000.0;

    stime =
	rusage.ru_stime.tv_sec + rusage.ru_stime.tv_usec / 1000000.0;

    ixrssSeconds = rusage.ru_ixrss / (clockTick + 0.0);
    idrssSeconds = rusage.ru_idrss / (clockTick + 0.0);
    isrssSeconds = rusage.ru_isrss / (clockTick + 0.0);

    if ((AppendNameDouble (interp, resultObj, "userTimeUsed", utime)
	  == TCL_ERROR)
      || (AppendNameDouble (interp, resultObj, "systemTimeUsed", stime) 
	  == TCL_ERROR)
      || (AppendNameLong (interp, resultObj, "maxResidentSetSizeKbytes", 
	    rusage.ru_maxrss) == TCL_ERROR)
      || (AppendNameDouble (interp, resultObj, "sharedTextKbyteSeconds", 
	    ixrssSeconds) == TCL_ERROR)
      || (AppendNameDouble (interp, resultObj, "unsharedDataKbyteSeconds", 
	    idrssSeconds) == TCL_ERROR)
      || (AppendNameDouble (interp, resultObj, "stackKbyteSeconds", 
	    isrssSeconds) == TCL_ERROR)
      || (AppendNameLong (interp, resultObj, "pageReclaims", 
	    rusage.ru_minflt) == TCL_ERROR)
      || (AppendNameLong (interp, resultObj, "pageFaults", 
	    rusage.ru_majflt) == TCL_ERROR)
      || (AppendNameLong (interp, resultObj, "nTimesSwapped", 
	    rusage.ru_nswap) == TCL_ERROR)
      || (AppendNameLong (interp, resultObj, "fileSystemInputs", 
	    rusage.ru_inblock) == TCL_ERROR)
      || (AppendNameLong (interp, resultObj, "fileSystemOutputs", 
	    rusage.ru_oublock) == TCL_ERROR)
      || (AppendNameLong (interp, resultObj, "ipcMessagesSent", 
	    rusage.ru_msgsnd) == TCL_ERROR)
      || (AppendNameLong (interp, resultObj, "ipcMessagesReceived", 
	    rusage.ru_msgrcv) == TCL_ERROR)
      || (AppendNameLong (interp, resultObj, "nSignalsDelivered", 
	    rusage.ru_nsignals) == TCL_ERROR)
      || (AppendNameLong (interp, resultObj, "voluntaryContextSwitches", 
	    rusage.ru_nvcsw) == TCL_ERROR)
      || (AppendNameLong (interp, resultObj, "involuntaryContextSwitches", 
	   rusage.ru_nivcsw) 
	  == TCL_ERROR))
	      return TCL_ERROR;

    return TCL_OK;
}

/*-----------------------------------------------------------------------------
 * BSD_RlimitObjCmd --
 *  
 * Implements the `rlimit' command:
 *    rlimit...
 *  
 * Results:
 *      A standard Tcl result.
 *      
 * Side effects:
 *      See the user documentation.
 *-----------------------------------------------------------------------------
 */     
int
BSD_RlimitObjCmd (clientData, interp, objc, objv)
    ClientData    clientData;
    Tcl_Interp   *interp;
    int           objc;
    Tcl_Obj      *CONST objv[];
{
    static CONST84 char *getSetOptions[] = { "get", "set", (char *) NULL};
    Tcl_WideInt limitVal = 0;

    enum GSSubOptIdx {
            GSGetIdx, GSSetIdx
    } getset;

    static CONST84 char *softHardOptions[] = { 
             "soft", "hard", "both",
             (char *) NULL};

    enum SHSubOptIdx {
            SHSoftIdx, SHHardIdx, SHBothIdx
    } hardsoft;

    static CONST84 char *limitOptions[] = { 
            "virtual", "core", "cpu", "data", "fsize",
             "memlock", "nofile", "nproc", "rss", 
             "sockbuf", "stack",
             (char *) NULL};

    enum ISubOptIdx {
            IVirtualMemoryIdx, ICoreIdx, ICpuIdx, IDataIdx, IFsizeIdx,
            IMemlockIdx, INofileIdx, INprocIdx, IRssIdx,
            ISockBufIdx, IStackIdx
    } limitID;

    Tcl_Obj        *resultObj = Tcl_GetObjResult (interp);
    int             result;
    int             resource = 0;
    struct rlimit   rlimit;

    if (objc < 4 || objc > 5) {
	Tcl_WrongNumArgs(interp, 1, objv, "get|set soft|hard limitName ?val?");
	return TCL_ERROR;
    }

    result = Tcl_GetIndexFromObj(interp, objv[1], getSetOptions, 
	        "get/set option", 0, (int *) &getset);
    if (result != TCL_OK) {
        return result;
    }
 
    result = Tcl_GetIndexFromObj(interp, objv[2], softHardOptions, 
	    "hard/soft option", 0, (int *) &hardsoft);
    if (result != TCL_OK) {
        return result;
    }
  
    result = Tcl_GetIndexFromObj(interp, objv[3], limitOptions, "limit name", 0,
            (int *) &limitID);
    if (result != TCL_OK) {
        return result;
    }
 
    switch (limitID) {
	case IVirtualMemoryIdx:
	    resource = RLIMIT_AS;
	    break;

        case ICoreIdx:
	    resource = RLIMIT_CORE;
	    break;

        case ICpuIdx:
	    resource = RLIMIT_CPU;
	    break;

        case IDataIdx:
	    resource = RLIMIT_DATA;
	    break;

        case IFsizeIdx:
	    resource = RLIMIT_FSIZE;
	    break;

        case IMemlockIdx:
	    resource = RLIMIT_MEMLOCK;
	    break;

        case INofileIdx:
	    resource = RLIMIT_NOFILE;
	    break;

        case INprocIdx:
	    resource = RLIMIT_NPROC;
	    break;

        case IRssIdx:
	    resource = RLIMIT_RSS;
	    break;

	case ISockBufIdx:
#ifdef RLIMIT_SBSIZE
	    resource = RLIMIT_SBSIZE;
	    break;
#else
	    Tcl_SetStringObj (resultObj, "sockbuf option is not available on this operating system", -1);
	    return TCL_ERROR;
#endif // RLIMIT_SBSIZE 

        case IStackIdx:
	    resource = RLIMIT_STACK;
	    break;
    }

    switch (getset) {

	static Tcl_Obj *unlimitedResultObj = NULL;

	case GSGetIdx:

	    if (unlimitedResultObj == NULL) {
		unlimitedResultObj = Tcl_NewStringObj ("unlimited", -1);
	    }

	    if (objc != 4) {
		Tcl_WrongNumArgs(interp, 1, objv, "get soft|hard limitName");
		return TCL_ERROR;
	    }

	    if (getrlimit (resource, &rlimit) < 0) {
		Tcl_SetStringObj (resultObj, Tcl_PosixError (interp), -1);
		return TCL_ERROR;
	    }
	   
            switch (hardsoft) {
		case SHSoftIdx:
		    limitVal = rlimit.rlim_cur;
		    break;

		case SHHardIdx:
		    limitVal = rlimit.rlim_max;
		    break;

		case SHBothIdx:
		    panic ("unexpected/impossible code path");
	    }

	    if (limitVal == RLIM_INFINITY) {
		if (Tcl_ListObjAppendElement (interp, resultObj, unlimitedResultObj) == TCL_ERROR) {
		    return TCL_ERROR;
		}
		Tcl_IncrRefCount (unlimitedResultObj);
	    } else {
		if (Tcl_ListObjAppendElement (interp, resultObj, Tcl_NewWideIntObj (limitVal)) == TCL_ERROR) {
		    return TCL_ERROR;
		}
	    }

	    break;

	case GSSetIdx:
	    if (objc != 5) {
		Tcl_WrongNumArgs(interp, 1, objv, 
		    "set soft|hard limitName value");
		return TCL_ERROR;
	    }

	    if (strcmp (Tcl_GetString (objv[4]), "unlimited") == 0) {
		limitVal = RLIM_INFINITY;
	    } else {
		if (Tcl_GetWideIntFromObj (interp, objv[4], &limitVal) 
		      == TCL_ERROR) {
		    return TCL_ERROR;
		}
	    }

	    if (getrlimit (resource, &rlimit) < 0) {
		Tcl_SetStringObj (resultObj, Tcl_PosixError (interp), -1);
		return TCL_ERROR;
	    }
	   
            switch (hardsoft) {
		case SHSoftIdx:
		    rlimit.rlim_cur = limitVal;
		    break;

		case SHHardIdx:
		    rlimit.rlim_max = limitVal;
		    break;

		case SHBothIdx:
		    panic ("unexpected/impossible code path");
	    }

	    if (setrlimit (resource, &rlimit) < 0) {
		Tcl_SetStringObj (resultObj, Tcl_PosixError (interp), -1);
		return TCL_ERROR;
	    }
	    break;
    }

    return TCL_OK;
}

/*-----------------------------------------------------------------------------
 * StatfsBufToList --
 *  
 * Appends the contents of a statfs buffer to a Tcl list as key-value pairs.
 *  
 * Results:
 *      A standard Tcl result.
 *      
 *-----------------------------------------------------------------------------
 */     
static int
StatfsBufToList (Tcl_Interp *interp, Tcl_Obj *listObj, struct statfs *sp)
{
    Tcl_Obj      *flagListObj = Tcl_NewObj ();

#ifdef MFSNAMELEN
    char          fstypename[MFSNAMELEN+1];
    char          mntonname[MNAMELEN+1];
    char          mntfromname[MNAMELEN+1];

    strncpy (fstypename, sp->f_fstypename, MFSNAMELEN);
    strncpy (mntonname, sp->f_mntonname, MNAMELEN);
    strncpy (mntfromname, sp->f_mntfromname, MNAMELEN);
#endif

    if ((AppendNameLong (interp, listObj, 
			 "fundamentalFileSystemBlockSize", sp->f_bsize)
	  == TCL_ERROR)
#ifdef HAVE_STRUCT_STATFS_F_IOSIZE
      || (AppendNameLong (interp, listObj, 
			  "optimalTransferBlockSize", sp->f_iosize) 
	  == TCL_ERROR)
#endif
      || (AppendNameLong (interp, listObj, 
			  "totalDataBlocks", sp->f_blocks)
	  == TCL_ERROR)
      || (AppendNameLong (interp, listObj, 
			  "freeBlocks", sp->f_bfree)
	  == TCL_ERROR)
      || (AppendNameLong (interp, listObj, 
			  "availableFreeBlocks", sp->f_bavail)
	  == TCL_ERROR)
      || (AppendNameLong (interp, listObj, 
			  "totalFileNodes", sp->f_files)
	  == TCL_ERROR)
      || (AppendNameLong (interp, listObj, 
			  "freeFileNodes", sp->f_ffree)
	  == TCL_ERROR)
#ifdef MFSNAMELEN
      || (AppendNameString (interp, listObj, 
			  "fileSystemType", fstypename)
	  == TCL_ERROR)
      || (AppendNameString (interp, listObj, 
			  "mountPoint", mntonname)
	  == TCL_ERROR)
      || (AppendNameString (interp, listObj, 
			  "mountedFileSystem", mntfromname)
	  == TCL_ERROR)
#endif
    )
	return TCL_ERROR;

#define FLAGCHECK(x, y) if (sp->f_flags & x) {if (Tcl_ListObjAppendElement (interp, flagListObj, Tcl_NewStringObj (y, -1)) != TCL_OK) return TCL_ERROR;}

#ifdef MNT_RDONLY
    FLAGCHECK (MNT_RDONLY, "readOnly")
#endif

#ifdef MNT_SYNCHRONOUS
    FLAGCHECK (MNT_SYNCHRONOUS, "synchronous")
#endif

#ifdef MNT_NOEXEC
    FLAGCHECK (MNT_NOEXEC, "noExec")
#endif

#ifdef MNT_NOSUID
    FLAGCHECK (MNT_NOSUID, "noSUID")
#endif

#ifdef MNT_NODEV
    FLAGCHECK (MNT_NODEV, "noDev")
#endif

#ifdef MNT_UNION
    FLAGCHECK (MNT_UNION, "union")
#endif

#ifdef MNT_ASYNC
    FLAGCHECK (MNT_ASYNC, "asynchronous")
#endif

#ifdef MNT_SUIDDIR
    FLAGCHECK (MNT_SUIDDIR, "SUIDdir")
#endif

#ifdef MNT_SOFTDEP
    FLAGCHECK (MNT_SOFTDEP, "softUpdates")
#endif

#ifdef MNT_GJOURNAL
    FLAGCHECK (MNT_GJOURNAL, "gjournal")
#endif

#ifdef MNT_NOSYMFOLLOW
    FLAGCHECK (MNT_NOSYMFOLLOW, "noFollowSymlinks")
#endif

#ifdef MNT_NOATIME
    FLAGCHECK (MNT_NOATIME, "noAtime")
#endif

#ifdef MNT_NOCLUSTERR
    FLAGCHECK (MNT_NOCLUSTERR, "noClusterRead")
#endif

#ifdef MNT_NOCLUSTERW
    FLAGCHECK (MNT_NOCLUSTERW, "noClusterWrite")
#endif

#ifdef MNT_EXRDONLY
    FLAGCHECK (MNT_EXRDONLY, "exportedReadOnly")
#endif

#ifdef MNT_EXPORTED
    FLAGCHECK (MNT_EXPORTED, "exported")
#endif

#ifdef MNT_DEFEXPORTED
    FLAGCHECK (MNT_DEFEXPORTED, "worldExported")
#endif

#ifdef MNT_EXPORTANON
    FLAGCHECK (MNT_EXPORTANON, "anonUidMapping")
#endif

#ifdef MNT_EXKERB
    FLAGCHECK (MNT_EXKERB, "kerberosUidMapping")
#endif

#ifdef MNT_EXPUBLIC
    FLAGCHECK (MNT_EXPUBLIC, "webNFS")
#endif

#ifdef MNT_LOCAL
    FLAGCHECK (MNT_LOCAL, "filesystemStoredLocally")
#endif

#ifdef MNT_QUOTA
    FLAGCHECK (MNT_QUOTA, "quotasEnabled")
#endif

#ifdef MNT_ROOTFS
    FLAGCHECK (MNT_ROOTFS, "rootFilesystem")
#endif

#ifdef MNT_USER
    FLAGCHECK (MNT_USER, "userMounted")
#endif

#ifdef MNT_MULTILABEL
    FLAGCHECK (MNT_MULTILABEL, "multiLabel")
#endif

#ifdef MNT_ACLS
    FLAGCHECK (MNT_ACLS, "aclsEnabled")
#endif

#ifdef MNT_JAILDEVFS
    FLAGCHECK (MNT_JAILDEVFS, "jailFriendlyDevfsBehavior")
#endif

#ifdef MNT_IGNORE
    FLAGCHECK (MNT_IGNORE, "doNotShowInDF")
#endif

    if (AppendNameObj (interp, listObj, "flags", flagListObj) == TCL_ERROR) return TCL_ERROR;

    return TCL_OK;
}

/*-----------------------------------------------------------------------------
 * BSD_StatfsObjCmd --
 *  
 * Implements the `statfs' command:
 *    statfs path
 *  
 * Results:
 *      A standard Tcl result.
 *      
 * Side effects:
 *      See the user documentation.
 *-----------------------------------------------------------------------------
 */     
int
BSD_StatfsObjCmd (clientData, interp, objc, objv)
    ClientData    clientData;
    Tcl_Interp   *interp;
    int           objc;
    Tcl_Obj      *CONST objv[];
{
    char         *path;
    Tcl_Obj      *resultObj = Tcl_GetObjResult (interp);
    struct statfs statfsbuf;

    if (objc != 2) {
	Tcl_WrongNumArgs(interp, 1, objv, "path");
	return TCL_ERROR;
    }

    path = Tcl_GetStringFromObj (objv[1], NULL);

    if (statfs (path, &statfsbuf) < 0) {
	Tcl_SetStringObj (resultObj, Tcl_PosixError (interp), -1);
	return TCL_ERROR;
    }

    return StatfsBufToList (interp, resultObj, &statfsbuf);
}

/*-----------------------------------------------------------------------------
 * BSD_GetfsstatObjCmd --
 *  
 * Implements the `getfsstat' command:
 *    getfsstat ?-wait|-nowait?
 *  
 * Results:
 *      A standard Tcl result.
 *      
 * Side effects:
 *      See the user documentation.
 *-----------------------------------------------------------------------------
 */     
int
BSD_GetfsstatObjCmd (clientData, interp, objc, objv)
    ClientData    clientData;
    Tcl_Interp   *interp;
    int           objc;
    Tcl_Obj      *CONST objv[];
{
#ifdef HAVE_GETFSSTAT
    Tcl_Obj       *resultObj = Tcl_GetObjResult (interp);
    struct statfs *statfsbufs;
    int            bufsize;
    int            i;
    int            nMountedFilesystems;
    int            waitOption = MNT_NOWAIT;
    char          *subOption;

    if (objc > 2) {
      usage:
	Tcl_WrongNumArgs(interp, 1, objv, "?-wait|-nowait?");
	return TCL_ERROR;
    }

    if (objc == 2) {
	subOption = Tcl_GetStringFromObj (objv[1], NULL);
	if (strcmp (subOption, "-nowait") == 0) {
	    waitOption = MNT_NOWAIT;
	} else if (strcmp (subOption, "-wait") == 0) {
	    waitOption = MNT_WAIT;
	} else {
	    goto usage;
	}
    }

    nMountedFilesystems = getfsstat (NULL, 0, waitOption);
    if (nMountedFilesystems < 0) {
	Tcl_SetStringObj (resultObj, Tcl_PosixError (interp), -1);
	return TCL_ERROR;
    }

    bufsize = nMountedFilesystems * sizeof (struct statfs);

    statfsbufs = (struct statfs *)ckalloc (bufsize);

    if (getfsstat (statfsbufs, bufsize, MNT_NOWAIT) < 0) {
	Tcl_SetStringObj (resultObj, Tcl_PosixError (interp), -1);
	ckfree ((void *)statfsbufs);
	return TCL_ERROR;
    }

    for (i = 0; i < nMountedFilesystems; i++) {
	Tcl_Obj *listObj = Tcl_NewObj ();

        if (StatfsBufToList (interp, listObj, &statfsbufs[i]) == TCL_ERROR) {
	    ckfree ((void *)statfsbufs);
	    return TCL_ERROR;
	}

	if (Tcl_ListObjAppendElement (interp, resultObj, listObj) == TCL_ERROR) {
	    ckfree ((void *)statfsbufs);
	    return TCL_ERROR;
	}
    }

    ckfree ((void *)statfsbufs);
    return TCL_OK;
#else // HAVE_GETFSSTAT
    Tcl_SetObjResult (interp, Tcl_NewStringObj ("getfsstat is not available on this operating system", -1));
    return TCL_ERROR;
#endif// HAVE_GETFSSTAT
}


/*-----------------------------------------------------------------------------
 * BSD_GetLoadAvgObjCmd --
 *  
 * Implements the `getloadavg' command:
 *    getloadavg
 *  
 * Results:
 *      A standard Tcl result.
 *      
 * Side effects:
 *      See the user documentation.
 *-----------------------------------------------------------------------------
 */     
int
BSD_GetLoadAvgObjCmd (clientData, interp, objc, objv)
    ClientData    clientData;
    Tcl_Interp   *interp;
    int           objc;
    Tcl_Obj      *CONST objv[];
{
#define LOADAVG_NELEM 3

    double        loadavg[LOADAVG_NELEM];
    Tcl_Obj      *loadavgObj;
    Tcl_Obj      *resultObj = Tcl_GetObjResult (interp);
    int           loadIndex;

    if (getloadavg (&loadavg[0], LOADAVG_NELEM) < 0) {
	Tcl_SetStringObj (resultObj, Tcl_PosixError (interp), -1);
	return TCL_ERROR;
    }

    for (loadIndex = 0; loadIndex < LOADAVG_NELEM; loadIndex++) {
	loadavgObj = Tcl_NewDoubleObj (loadavg[loadIndex]);
	if (Tcl_ListObjAppendElement (interp, resultObj, loadavgObj) != TCL_OK) {
	    return TCL_ERROR;
	}
    }

    return TCL_OK;
}


/*-----------------------------------------------------------------------------
 * BSD_SetProcTitleObjCmd --
 *  
 * Implements the `setproctitle' command:
 *    setproctitle
 *  
 * Results:
 *      A standard Tcl result.
 *      
 * Side effects:
 *      See the user documentation.
 *-----------------------------------------------------------------------------
 */     
int
BSD_SetProcTitleObjCmd (clientData, interp, objc, objv)
    ClientData    clientData;
    Tcl_Interp   *interp;
    int           objc;
    Tcl_Obj      *CONST objv[];
{
    char *titleString;

    if (objc > 2) {
	Tcl_WrongNumArgs (interp, 1, objv, "[string]");
	return TCL_ERROR;
    }

#ifdef HAVE_SETPROCTITLE
    if (objc == 1) {
	titleString = NULL;
    } else {
	titleString = Tcl_GetString (objv[1]);
    }

    setproctitle ("-%s", titleString);
    return TCL_OK;
#else
    Tcl_SetObjResult (interp, Tcl_NewStringObj ("setproctitle is not available on this operating system", -1));
    return TCL_ERROR;
#endif
}

/*-----------------------------------------------------------------------------
 * BSD_AbortCmd --
 *  
 * Implements the `abort' command:
 *    abort
 *  
 * Results:
 *      An abort signal is generated and probably aborts Tcl.
 *
 * Side effects:
 *      See the user documentation.
 *-----------------------------------------------------------------------------
 */     
int
BSD_AbortCmd (clientData, interp, objc, objv)
    ClientData    clientData;
    Tcl_Interp   *interp;
    int           objc;
    Tcl_Obj      *CONST objv[];
{

    abort();

    return TCL_OK;
}
